### PR TITLE
DCMAW-9802: Refactor of Client Credential service

### PR DESCRIPTION
**Pull request by sandy-may**
​DCMAW-9802
### What changed
Refactor of the client credential service:
- Renamed to client registry service
- Expose two public methods on the class for retrieving registered client details

### Why did it change
- There is data in the client registry that doesn't relate to client credentials
- The lambdas had to orchestrate the service adding extra complexity to the lambda handler. This simplifies it so the lambda only knows what it needs to know to perform its job.

## Checklists

- [x] There is a ticket raised for this PR that is present in the branch name
- [x] No PII data logged. [See guidance here](https://govukverify.atlassian.net/wiki/spaces/DCMAW/pages/3502407722/PII+Logging+Considerations)
- [ ] Demo to a BA, TA, and the team.
- [ ] Update [README](./blob/main/README.md) with any new instructions or tasks

### CODE REVIEW

File: backend-api/src/functions/services/clientRegistryService/clientRegistryService.ts [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+export class ClientRegistryService
+  implements
+    IGetRegisteredIssuerUsingClientSecrets,
+    IGetPartialRegisteredClientByClientId
+{
+  ssmClient: SSMClient;
+  cacheTTL: number;
+
+  constructor() {
+    this.ssmClient = new SSMClient(ssmClient);
```
Comment (J-son1) [inline backend-api/src/functions/services/clientRegistryService/clientRegistryService.ts:21]: I think the `ssmClient` argument here is already an instance of a `SSMClient`
```suggestion
    this.ssmClient = ssmClient;
```
Also, should the ssmClient be passed through the constructor as a default parameter?

Comment (sandy-may) [reply backend-api/src/functions/services/clientRegistryService/clientRegistryService.ts]: Existing issue but updated here :) Thanks for raising

File: backend-api/src/functions/services/clientRegistryService/clientRegistryService.ts [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+{
+  ssmClient: SSMClient;
+  cacheTTL: number;
+
+  constructor() {
+    this.ssmClient = ssmClient;
+    this.cacheTTL = 3600 * 1000;
+  }
+  getRegisteredIssuerUsingClientSecrets = async (credentials: {
+    clientId: string;
```
Comment (sandy-may) [inline backend-api/src/functions/services/clientRegistryService/clientRegistryService.ts:25]: use the decoded client credential interface?

File: backend-api/src/functions/services/clientRegistryService/clientRegistryService.ts [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    );
+    if (!registeredClient) return errorResult("Client is not registered");
+
+    return successResult({
+      issuer: registeredClient.issuer,
+      redirectUri: registeredClient.redirect_uri,
+    });
+  };
+
+  private getClientRegistery = async (): Promise<Result<IClientRegistry>> => {
```
Comment (sandy-may) [inline backend-api/src/functions/services/clientRegistryService/clientRegistryService.ts:72]: spelling

File: backend-api/src/functions/services/clientRegistryService/clientRegistryService.ts [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    if (cache && cache.expiry > Date.now()) {
+      return successResult(cache.data);
+    }
+    const command: GetParameterRequest = {
+      Name: "/dev/async-credential/CLIENT_CREDENTIALS",
```
Comment (sandy-may) [inline backend-api/src/functions/services/clientRegistryService/clientRegistryService.ts:77]: This has to be parameterised - suggest using an environment variable

File: backend-api/src/functions/services/clientRegistryService/clientRegistryService.ts [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    incomingClientSecrets: { clientId: string; clientSecret: string },
+  ): boolean => {
+    const storedSalt = registeredClientSecrets.salt;
+    const hashedSuppliedClientSecret = hashSecret(
+      incomingClientSecrets.clientSecret,
+      storedSalt,
+    );
+    const hashedStoredClientSecret = registeredClientSecrets.hashedClientSecret;
+
+    console.log("REGISTERED CLIENT SECRET", hashedStoredClientSecret);
```
Comment (sandy-may) [inline backend-api/src/functions/services/clientRegistryService/clientRegistryService.ts:156]: remove logs!

File: backend-api/src/functions/services/clientRegistryService/clientRegistryService.ts (hunk 2 above)
Comment (sandy-may) [reply backend-api/src/functions/services/clientRegistryService/clientRegistryService.ts]: done

File: backend-api/src/functions/services/clientRegistryService/clientRegistryService.ts (hunk 3 above)
Comment (sandy-may) [reply backend-api/src/functions/services/clientRegistryService/clientRegistryService.ts]: done

File: backend-api/src/functions/services/clientRegistryService/clientRegistryService.ts (hunk 5 above)
Comment (sandy-may) [reply backend-api/src/functions/services/clientRegistryService/clientRegistryService.ts]: done

**Review by J-son1 — APPROVED** [review]

File: backend-api/src/functions/services/clientRegistryService/clientRegistryService.ts (hunk 4 above)
Comment (sandy-may) [reply backend-api/src/functions/services/clientRegistryService/clientRegistryService.ts]: done

**Pull request merged**